image/image_linux.go

```diff
@@ -272,302 +273,229 @@ func makeLabel(now time.Time) string {
 	return now.UTC().Format("20060102")
 }
 
-func selectAssertionMaxFormats(tsto *tooling.ToolingStore, sysSn *seedwriter.SeedSnap) error {
-	if sysSn == nil {
-		// nothing to do
-		return nil
+type imageSeeder struct {
+	classic        bool
+	prepareDir     string
+	wideCohortKey  string
+	revisions      map[string]snap.Revision
+	customizations *Customizations
+
+	hasModes    bool
+	rootDir     string
+	bootRootDir string
+	seedDir     string
+	label       string
+
+	model *asserts.Model
+	w     *seedwriter.Writer
+	tsto  *tooling.ToolingStore
+}
+
+func (s *imageSeeder) setModelessDirs() error {
+	if s.classic {
+		// Classic, PrepareDir is the root dir itself
+		s.rootDir = s.prepareDir
+	} else {
+		// Core 16/18,  writing for the writeable partition
+		s.rootDir = filepath.Join(s.prepareDir, "image")
+		s.bootRootDir = s.rootDir
 	}
-	snapf, err := snapfile.Open(sysSn.Path)
-	if err != nil {
-		return err
+	s.seedDir = dirs.SnapSeedDirUnder(s.rootDir)
+
+	// validity check target
+	if osutil.FileExists(dirs.SnapStateFileUnder(s.rootDir)) {
+		return fmt.Errorf("cannot prepare seed over existing system or an already booted image, detected state file %s", dirs.SnapStateFileUnder(s.rootDir))
 	}
-	// XXX take also kernel into account
-	// XXX warning logic
-	maxFormats, _, err := snap.SnapdAssertionMaxFormatsFromSnapFile(snapf)
-	if err != nil {
-		return err
+	if snaps, _ := filepath.Glob(filepath.Join(dirs.SnapBlobDirUnder(s.rootDir), "*.snap")); len(snaps) > 0 {
+		return fmt.Errorf("expected empty snap dir in rootdir, got: %v", snaps)
 	}
-	tsto.SetAssertionMaxFormats(maxFormats)
 	return nil
 }
 
-var setupSeed = func(tsto *tooling.ToolingStore, model *asserts.Model, opts *Options) error {
-	if model.Classic() != opts.Classic {
-		return fmt.Errorf("internal error: classic model but classic mode not set")
+func (s *imageSeeder) setModesDirs() error {
+	// Core 20, writing for the system-seed partition
+	s.seedDir = filepath.Join(s.prepareDir, "system-seed")
+	s.label = makeLabel(time.Now())
+	s.bootRootDir = s.seedDir
+
+	// validity check target
+	if systems, _ := filepath.Glob(filepath.Join(s.seedDir, "systems", "*")); len(systems) > 0 {
+		return fmt.Errorf("expected empty systems dir in system-seed, got: %v", systems)
 	}
+	return nil
+}
 
-	hasModes := model.Grade() != asserts.ModelGradeUnset
-	var rootDir string
-	var bootRootDir string
-	var seedDir string
-	var label string
-	if !hasModes {
-		if opts.Classic {
-			// Classic, PrepareDir is the root dir itself
-			rootDir = opts.PrepareDir
-		} else {
-			// Core 16/18,  writing for the writeable partition
-			rootDir = filepath.Join(opts.PrepareDir, "image")
-			bootRootDir = rootDir
-		}
-		seedDir = dirs.SnapSeedDirUnder(rootDir)
+func newImageSeeder(tsto *tooling.ToolingStore, model *asserts.Model, opts *Options) (*imageSeeder, error) {
```

Comment:
nitpick: usually new* methods are defined immediately after the struct

Reply:
Moved `newImageSeeder` directly under struct